### ISSUE TITLE
`dict_eval` does not take into account support of elements. 

### ISSUE THREAD

**Issue by vincentcp** [JuliaApproximation/BasisFunctions.jl#6]
`dict_eval`  at https://github.com/JuliaApproximation/BasisFunctions.jl/blob/9b31b5350949c56dc9bbf94fb231b67e2216515b/src/bases/dictionary/dict_evaluation.jl#L207 only checks for dictionary support. Not for dictionary element support. If one element of the dictionary has support `s` smaller the dictionary support `S` evaluation at `x∈ S\s` will be incorrect.

**Comment by vincentcp** [comment]
`default_dict_innerproduct`  at 
https://github.com/JuliaApproximation/BasisFunctions.jl/blob/9b31b5350949c56dc9bbf94fb231b67e2216515b/src/bases/dictionary/gram.jl#L55 
uses `unsafe_eval_element(Φ1, i, x)` that again leads to incorrect evaluation if dictionary elements have smaller support than the dictionary itself.

**Comment by daanhb** [comment]
I feel a version 0.10.1 coming up shortly :-)

In the first case, the default iterator should indeed call `eval_element`, or perhaps `unsafe_eval_element1`. Concrete dictionaries can return something more efficient because they know more, e.g., the three-term-recurrence iterator of the orthogonal polynomials can do a support check only once.

In the second case things are more complicated, because the given measure may have a support that is wider than the support of the basis functions. Yet, the support check is possibly more expensive here, being part of the adaptive computation of an integral. There is a similar case in the `analysis_integral` function in basisfunction.jl, involving only one basis function. There, the integral is computed on the intersection of the supports of the measure and the basis function. This took some time to get right (and it may still not be right in all cases), because the computation of the intersection of two domains does not always succeed.
An additional problem here with compactly supported functions is that the integral may be evaluated incorrectly if the integrand is zero except in a small interval. So simply using `eval_element` with boundscheck may not be sufficient, and truncating the integration domain is preferrable.

**Comment by vincentcp** [comment]
The second case indeed prevents that the test in https://github.com/FrameFunVC/CompactTranslatesDict.jl/blob/2cb5615e4532b2062fcb5b2eb28ac16a67ad2a0e/test/runtests.jl#L44 is still consistent. `BasisFunctions.default_gram(B, μ)` returns the wrong Gram matrix because of the implicit assumption made in `default_dict_innerproduct`. 

To you suggest to add the logic in `analysis_integral` also to `default_dict_innerproduct`?

**Comment by daanhb** [comment]
Something along those lines, yes, with a fallback in case the intersection only results in an IntersectionDomain.

**Comment by vincentcp** [comment]
Should the macro @inbounds be added ?
``` julia
function default_dict_innerproduct(Φ1::Dictionary, i, Φ2::Dictionary, j, measure; options...)
  @boundscheck checkbounds(Φ1, i)
  @boundscheck checkbounds(Φ2, i)
       
  @inbounds applymeasure(measure, x->conj(unsafe_eval_element(Φ1, i, x)) * unsafe_eval_element(Φ2, j, x); options...)
end
```

**Comment by vincentcp** [comment]
I believe that this default is also efficient, but returns the wrong value because of the unsafe_eval, or does it?
https://github.com/JuliaApproximation/BasisFunctions.jl/blob/9b31b5350949c56dc9bbf94fb231b67e2216515b/src/bases/dictionary/basisfunction.jl#L78

**Comment by vincentcp** [comment]
Fixed by #8
